authz/providers/azure/rbac/checkaccessreqhelper.go

```diff
@@ -0,0 +1,234 @@
+	ConditionVersion        string `json:"ConditionVersion"`
+}
+type AzureDenyAssignment struct {
+	IsSystemProtected string `json:"IsSystemProtected"`
+	DenyAssignment
+}
+
+type AuthorizationDecesion struct {
+	Decesion            string              `json:"accessDecision"`
+	ActionId            string              `json:"actionId"`
+	IsDataAction        bool                `json:"isDataAction"`
+	AzureRoleAssignment AzureRoleAssignment `json:"roleAssignment"`
+	AzureDenyAssignment AzureDenyAssignment `json:"denyAssignment"`
+	TimeToLiveInMs      int                 `json:"timeToLiveInMs"`
+}
+
+// this will be removed once caching is implemented
+func getUserId(userName string) string {
+	switch userName {
+	case "krdhruva@microsoft.com":
+		return "63e8a863-9ae9-4f3c-b0b7-fd9df05c712e"
+	case "test@KDOrg.onmicrosoft.com":
+		return "62103f2e-051d-48cc-af47-b1ff3deec630"
+	default:
+		return "62103f2e-051d-48cc-af47-b1ff3deec630"
+	}
+}
+
+func getScope(resourceId string, attr *authzv1.ResourceAttributes) string {
+	if attr != nil && attr.Namespace != "" {
+		return resourceId + "/namespace/" + attr.Namespace
+	}
+	return resourceId
+}
+
+func IsValidUUID(u string) bool {
+	_, err := uuid.Parse(u)
+	return err == nil
+}
+
+func getSecGroups(groups []string) []string {
+	var finalGroups []string
+	for _, element := range groups {
+		if IsValidUUID(element) {
+			finalGroups = append(finalGroups, element)
+		}
+	}
+	return finalGroups
+}
+
+func getActionName(verb string) string {
+	switch verb {
+	case "get":
+		fallthrough
+	case "list":
+		fallthrough
+	case "watch":
+		return "read"
+	case "create":
+		return "action"
+	case "update":
+		return "write"
+	case "delete":
+		return "delete"
+	default:
+		return ""
+	}
+}
+
+func getDataAction(subRevReq *authzv1.SubjectAccessReviewSpec, clusterType string) AuthorizationActionInfo {
+	authInfo := AuthorizationActionInfo{
+		IsDataAction: true}
+
+	authInfo.AuthorizationEntity.Id = clusterType
+	if subRevReq.ResourceAttributes != nil {
+		if subRevReq.ResourceAttributes.Group != "" {
+			authInfo.AuthorizationEntity.Id += "/" + subRevReq.ResourceAttributes.Group
+		}
+		authInfo.AuthorizationEntity.Id += "/" + subRevReq.ResourceAttributes.Resource + "/" + getActionName(subRevReq.ResourceAttributes.Verb)
+	} else if subRevReq.NonResourceAttributes != nil {
+		authInfo.AuthorizationEntity.Id += subRevReq.NonResourceAttributes.Path + "/" + getActionName(subRevReq.NonResourceAttributes.Verb)
+	}
+	return authInfo
+}
+
+func PrepareCheckAccessRequest(req *authzv1.SubjectAccessReviewSpec, clusterType, resourceId string) *CheckAccessRequest {
+	checkaccessreq := CheckAccessRequest{}
+	checkaccessreq.Subject.Attributes.ObjectId = getUserId(req.User)
+
+	if req.Groups != nil && len(req.Groups) > 0 {
+		groups := getSecGroups(req.Groups)
+		if groups != nil && len(groups) > 0 {
+			checkaccessreq.Subject.Attributes.Groups = groups
+			checkaccessreq.Subject.Attributes.ExpandGroupMembership = true
```

Comment:
what does ExpandGroupMembership mean?

Reply:
It is required if we don't pass any gropu list. Passing it true means, PAS will retrieve group info which will have 4 times more latency. As webhook is passing group info, this is not required. Will remove it in next iteration.